akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,687 @@
+
+  final class Firehose(
+      val firehoseKey: FirehoseKey,
+      val settings: Settings,
+      val firehoseHub: Source[EventEnvelope[Any], NotUsed],
+      firehoseKillSwitch: KillSwitch,
+      log: LoggingAdapter) {
+
+    val consumerTracking: ConcurrentHashMap[String, ConsumerTracking] = new ConcurrentHashMap
+    @volatile private var firehoseIsShutdown = false
+
+    private def entityType = firehoseKey.entityType
+    private val sliceRangeStr = s"${firehoseKey.sliceRange.min}-${firehoseKey.sliceRange.max}"
+
+    def consumerStarted(consumerId: String, consumerKillSwitch: KillSwitch): Unit = {
+      log.debug("Firehose entityType [{}] sliceRange [{}] consumer [{}] started", entityType, sliceRangeStr, consumerId)
+      consumerTracking.putIfAbsent(
+        consumerId,
+        ConsumerTracking(consumerId, Instant.EPOCH, firehoseOnly = false, consumerKillSwitch, None))
+    }
+
+    def consumerTerminated(consumerId: String): Int = {
+      log.debug(
+        "Firehose entityType [{}] sliceRange [{}] consumer [{}] terminated",
+        entityType,
+        sliceRangeStr,
+        consumerId)
+      consumerTracking.remove(consumerId)
+      consumerTracking.size
+    }
+
+    def shutdownFirehoseIfNoConsumers(): Boolean = {
+      if (consumerTracking.isEmpty) {
+        log.debug("Firehose entityType [{}] sliceRange [{}] is shutting down, no consumers", entityType, sliceRangeStr)
+        firehoseIsShutdown = true
+        firehoseKillSwitch.shutdown()
+        true
+      } else
+        false
+    }
+
+    def isShutdown: Boolean =
+      firehoseIsShutdown
+
+    @tailrec def updateConsumerTracking(
+        consumerId: String,
+        timestamp: Instant,
+        consumerKillSwitch: KillSwitch): Unit = {
+
+      val existingTracking = consumerTracking.get(consumerId)
+      val tracking = existingTracking match {
+        case null =>
+          ConsumerTracking(consumerId, timestamp, firehoseOnly = false, consumerKillSwitch, None)
+        case existing =>
+          if (timestamp.isAfter(existing.timestamp))
+            existing.copy(timestamp = timestamp)
+          else
+            existing
+      }
+
+      if (!consumerTracking.replace(consumerId, existingTracking, tracking)) {
+        // concurrent update, try again
+        updateConsumerTracking(consumerId, timestamp, consumerKillSwitch)
+      }
+    }
+
+    def detectSlowConsumers(now: Instant): Unit = {
+      import akka.util.ccompat.JavaConverters._
+      val consumerTrackingValues = consumerTracking.values.iterator.asScala.toVector
+      if (consumerTrackingValues.size > 1) {
+        val slowestConsumer = consumerTrackingValues.minBy(_.timestamp)
+        val fastestConsumer = consumerTrackingValues.maxBy(_.timestamp)
+
+        val slowConsumers = consumerTrackingValues.collect {
+          case t
+              if t.firehoseOnly &&
+              isDurationGreaterThan(t.timestamp, fastestConsumer.timestamp, settings.slowConsumerBehindThreshold) =>
```

Comment:
I realized that this will not work as intended. The buffer size of the BroadcastHub will slow down the fastest so that the timestamp diff will not increase. I'm working on refined logic.

Reply:
Third time is the charm.